Problem3.java

```diff
@@ -0,0 +1,44 @@
+/**
+ * Definition for singly-linked list.
+ * class ListNode {
+ *     int val;
+ *     ListNode next;
+ *     ListNode(int x) {
+ *         val = x;
+ *         next = null;
+ *     }
+ * }
+ */
+/**
+Leet Code Submitted : YES
+Space Complexity : O(N)
+Time Complexity : O(N) 
+
+The idea is to use HashSet to store the previous found node and then check if the node already exists. If yes then thats the cycle.
+**/
+
+public class Solution {
+    public ListNode detectCycle(ListNode head) {
+        Set<ListNode> s = new HashSet<>();
+        int flag = 0;
+        
```

Comment:
Try to optimize the solution

Reply:
Yes, I am working on optimizing it

Reply:
Added the new code with optimized solution

Thanks